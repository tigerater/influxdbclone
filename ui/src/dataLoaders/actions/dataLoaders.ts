// Libraries
import {get} from 'lodash'
import {normalize} from 'normalizr'

// APIs
import {client} from 'src/utils/api'
import {ScraperTargetRequest, PermissionResource} from '@influxdata/influx'
import {createAuthorization} from 'src/authorizations/apis'
import {postWrite as apiPostWrite, postLabel as apiPostLabel} from 'src/client'

// Schemas
import * as schemas from 'src/schemas'

// Utils
import {createNewPlugin} from 'src/dataLoaders/utils/pluginConfigs'
import {addLabelDefaults} from 'src/labels/utils'
import {getDataLoaders, getSteps} from 'src/dataLoaders/selectors'
import {getOrg} from 'src/organizations/selectors'

// Constants
import {
  pluginsByBundle,
  telegrafPluginsInfo,
} from 'src/dataLoaders/constants/pluginConfigs'

// Types
import {
  TelegrafPlugin,
  TelegrafPluginName,
  DataLoaderType,
  LineProtocolTab,
  Plugin,
  BundleName,
  ConfigurationState,
} from 'src/types/dataLoaders'
import {
  GetState,
  RemoteDataState,
  LabelProperties,
  Authorization,
  AuthEntities,
} from 'src/types'
import {ILabel} from '@influxdata/influx'
import {
  WritePrecision,
  TelegrafRequest,
  TelegrafPluginOutputInfluxDBV2,
  Permission,
} from '@influxdata/influx'
import {Dispatch} from 'redux'
import {addTelegraf, editTelegraf} from 'src/telegrafs/actions'
import {addAuthorization} from 'src/authorizations/actions/creators'
import {notify} from 'src/shared/actions/notifications'
import {
  TelegrafConfigCreationError,
  TelegrafConfigCreationSuccess,
  readWriteCardinalityLimitReached,
} from 'src/shared/copy/notifications'

const DEFAULT_COLLECTION_INTERVAL = 10000

export type Action =
  | SetDataLoadersType
  | SetTelegrafConfigID
  | UpdateTelegrafPluginConfig
  | AddConfigValue
  | RemoveConfigValue
  | SetActiveTelegrafPlugin
  | SetLineProtocolBody
  | SetActiveLPTab
  | SetLPStatus
  | SetPrecision
  | UpdateTelegrafPlugin
  | AddPluginBundle
  | AddTelegrafPlugins
  | RemoveBundlePlugins
  | RemovePluginBundle
  | SetPluginConfiguration
  | SetConfigArrayValue
  | SetScraperTargetBucket
  | SetScraperTargetURL
  | SetScraperTargetName
  | SetScraperTargetID
  | ClearDataLoaders
  | SetTelegrafConfigName
  | SetTelegrafConfigDescription
  | SetToken

interface SetDataLoadersType {
  type: 'SET_DATA_LOADERS_TYPE'
  payload: {type: DataLoaderType}
}

export const setDataLoadersType = (
  type: DataLoaderType
): SetDataLoadersType => ({
  type: 'SET_DATA_LOADERS_TYPE',
  payload: {type},
})

interface ClearDataLoaders {
  type: 'CLEAR_DATA_LOADERS'
}

export const clearDataLoaders = (): ClearDataLoaders => ({
  type: 'CLEAR_DATA_LOADERS',
})

interface SetTelegrafConfigName {
  type: 'SET_TELEGRAF_CONFIG_NAME'
  payload: {name: string}
}

export const setTelegrafConfigName = (name: string): SetTelegrafConfigName => ({
  type: 'SET_TELEGRAF_CONFIG_NAME',
  payload: {name},
})

interface SetTelegrafConfigDescription {
  type: 'SET_TELEGRAF_CONFIG_DESCRIPTION'
  payload: {description: string}
}

export const setTelegrafConfigDescription = (
  description: string
): SetTelegrafConfigDescription => ({
  type: 'SET_TELEGRAF_CONFIG_DESCRIPTION',
  payload: {description},
})

interface UpdateTelegrafPluginConfig {
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG'
  payload: {name: string; field: string; value: string}
}

export const updateTelegrafPluginConfig = (
  name: string,
  field: string,
  value: string
): UpdateTelegrafPluginConfig => ({
  type: 'UPDATE_TELEGRAF_PLUGIN_CONFIG',
  payload: {name, field, value},
})

interface UpdateTelegrafPlugin {
  type: 'UPDATE_TELEGRAF_PLUGIN'
  payload: {plugin: Plugin}
}

export const updateTelegrafPlugin = (plugin: Plugin): UpdateTelegrafPlugin => ({
  type: 'UPDATE_TELEGRAF_PLUGIN',
  payload: {plugin},
})

interface AddConfigValue {
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const addConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): AddConfigValue => ({
  type: 'ADD_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface RemoveConfigValue {
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: string
    fieldName: string
    value: string
  }
}

export const removeConfigValue = (
  pluginName: string,
  fieldName: string,
  value: string
): RemoveConfigValue => ({
  type: 'REMOVE_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, fieldName, value},
})

interface SetConfigArrayValue {
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE'
  payload: {
    pluginName: TelegrafPluginName
    field: string
    valueIndex: number
    value: string
  }
}

export const setConfigArrayValue = (
  pluginName: TelegrafPluginName,
  field: string,
  valueIndex: number,
  value: string
): SetConfigArrayValue => ({
  type: 'SET_TELEGRAF_PLUGIN_CONFIG_VALUE',
  payload: {pluginName, field, valueIndex, value},
})

interface SetTelegrafConfigID {
  type: 'SET_TELEGRAF_CONFIG_ID'
  payload: {id: string}
}

export const setTelegrafConfigID = (id: string): SetTelegrafConfigID => ({
  type: 'SET_TELEGRAF_CONFIG_ID',
  payload: {id},
})

interface AddPluginBundle {
  type: 'ADD_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const addPluginBundle = (bundle: BundleName): AddPluginBundle => ({
  type: 'ADD_PLUGIN_BUNDLE',
  payload: {bundle},
})

interface RemovePluginBundle {
  type: 'REMOVE_PLUGIN_BUNDLE'
  payload: {bundle: BundleName}
}

export const removePluginBundle = (bundle: BundleName): RemovePluginBundle => ({
  type: 'REMOVE_PLUGIN_BUNDLE',
  payload: {bundle},
})
interface AddTelegrafPlugins {
  type: 'ADD_TELEGRAF_PLUGINS'
  payload: {telegrafPlugins: TelegrafPlugin[]}
}

export const addTelegrafPlugins = (
  telegrafPlugins: TelegrafPlugin[]
): AddTelegrafPlugins => ({
  type: 'ADD_TELEGRAF_PLUGINS',
  payload: {telegrafPlugins},
})

interface RemoveBundlePlugins {
  type: 'REMOVE_BUNDLE_PLUGINS'
  payload: {bundle: BundleName}
}

export const removeBundlePlugins = (
  bundle: BundleName
): RemoveBundlePlugins => ({
  type: 'REMOVE_BUNDLE_PLUGINS',
  payload: {bundle},
})

interface SetScraperTargetBucket {
  type: 'SET_SCRAPER_TARGET_BUCKET'
  payload: {bucket: string}
}

export const setScraperTargetBucket = (
  bucket: string
): SetScraperTargetBucket => ({
  type: 'SET_SCRAPER_TARGET_BUCKET',
  payload: {bucket},
})

interface SetScraperTargetURL {
  type: 'SET_SCRAPER_TARGET_URL'
  payload: {url: string}
}

export const setScraperTargetURL = (url: string): SetScraperTargetURL => ({
  type: 'SET_SCRAPER_TARGET_URL',
  payload: {url},
})

interface SetScraperTargetName {
  type: 'SET_SCRAPER_TARGET_NAME'
  payload: {name: string}
}

export const setScraperTargetName = (name: string): SetScraperTargetName => ({
  type: 'SET_SCRAPER_TARGET_NAME',
  payload: {name},
})

interface SetScraperTargetID {
  type: 'SET_SCRAPER_TARGET_ID'
  payload: {id: string}
}

export const setScraperTargetID = (id: string): SetScraperTargetID => ({
  type: 'SET_SCRAPER_TARGET_ID',
  payload: {id},
})

interface SetToken {
  type: 'SET_TOKEN'
  payload: {token: string}
}

export const setToken = (token: string): SetToken => ({
  type: 'SET_TOKEN',
  payload: {token},
})

export const addPluginBundleWithPlugins = (bundle: BundleName) => dispatch => {
  dispatch(addPluginBundle(bundle))
  const plugins = pluginsByBundle[bundle]
  dispatch(
    addTelegrafPlugins(
      plugins.map(p => {
        const isConfigured = !!telegrafPluginsInfo[p].fields
          ? ConfigurationState.Unconfigured
          : ConfigurationState.Configured

        return {
          name: p,
          active: false,
          configured: isConfigured,
          templateID: telegrafPluginsInfo[p].templateID,
        }
      })
    )
  )
}

export const removePluginBundleWithPlugins = (
  bundle: BundleName
) => dispatch => {
  dispatch(removePluginBundle(bundle))
  dispatch(removeBundlePlugins(bundle))
}

export const createOrUpdateTelegrafConfigAsync = () => async (
  dispatch,
  getState: GetState
) => {
  const {
    telegrafPlugins,
    telegrafConfigID,
    telegrafConfigName,
    telegrafConfigDescription,
  } = getDataLoaders(getState())
  const {name} = getOrg(getState())
  const {bucket} = getSteps(getState())

  const influxDB2Out = {
    name: TelegrafPluginOutputInfluxDBV2.NameEnum.InfluxdbV2,
    type: TelegrafPluginOutputInfluxDBV2.TypeEnum.Output,
    config: {
      urls: [`${window.location.origin}`],
      token: '$INFLUX_TOKEN',
      organization: name,
      bucket,
    },
  }

  const plugins = telegrafPlugins.reduce(
    (acc, tp) => {
      if (tp.configured === ConfigurationState.Configured) {
        return [...acc, tp.plugin || createNewPlugin(tp.name)]
      }

      return acc
    },
    [influxDB2Out]
  )

  if (telegrafConfigID) {
    const telegraf = await client.telegrafConfigs.update(telegrafConfigID, {
      name: telegrafConfigName,
      description: telegrafConfigDescription,
      plugins,
    })
    dispatch(editTelegraf(telegraf))
    dispatch(setTelegrafConfigID(telegrafConfigID))
    return
  }

  createTelegraf(dispatch, getState, plugins)
}

const createTelegraf = async (dispatch, getState: GetState, plugins) => {
  try {
    const state = getState()
    const {telegrafConfigName, telegrafConfigDescription} = getDataLoaders(
      state
    )
    const {bucket, bucketID} = getSteps(state)
    const org = getOrg(getState())

    const telegrafRequest: TelegrafRequest = {
      name: telegrafConfigName,
      description: telegrafConfigDescription,
      agent: {collectionInterval: DEFAULT_COLLECTION_INTERVAL},
      orgID: org.id,
      plugins,
    }

    // create telegraf config
    const tc = await client.telegrafConfigs.create(telegrafRequest)

    const permissions = [
      {
        action: Permission.ActionEnum.Write,
        resource: {
          type: PermissionResource.TypeEnum.Buckets,
          id: bucketID,
          orgID: org.id,
        },
      },
      {
        action: Permission.ActionEnum.Read,
        resource: {
          type: PermissionResource.TypeEnum.Telegrafs,
          id: tc.id,
          orgID: org.id,
        },
      },
    ]

    const token = {
      name: `${telegrafConfigName} token`,
      orgID: org.id,
      description: `WRITE ${bucket} bucket / READ ${telegrafConfigName} telegraf config`,
      permissions,
    }

    // create token
    const createdToken = await createAuthorization(token)

    // add token to data loader state
    dispatch(setToken(createdToken.token))

    const normAuth = normalize<Authorization, AuthEntities, string>(
      createdToken,
      schemas.auth
    )

    // add token to authorizations state
    dispatch(addAuthorization(normAuth))

    // create token label
    const properties = {
      color: '#FFFFFF',
      description: `token for telegraf config: ${telegrafConfigName}`,
      tokenID: createdToken.id,
    } as LabelProperties // hack to make compiler work

    const resp = await apiPostLabel({
      data: {
        orgID: org.id,
        name: `@influxdata.token-${new Date().getTime()}`, // fix for https://github.com/influxdata/influxdb/issues/15730
        properties,
      },
    })

    if (resp.status !== 201) {
      throw new Error(resp.data.message)
    }

    const createdLabel = addLabelDefaults(resp.data.label)

    // add label to telegraf config
    const label = await client.telegrafConfigs.addLabel(
      tc.id,
      createdLabel as ILabel
    )

    const config = {
      ...tc,
      labels: [label],
    }

    dispatch(setTelegrafConfigID(tc.id))
    dispatch(addTelegraf(config))
    dispatch(notify(TelegrafConfigCreationSuccess))
  } catch (error) {
    console.error(error.message)
    dispatch(notify(TelegrafConfigCreationError))
  }
}

interface SetActiveTelegrafPlugin {
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN'
  payload: {telegrafPlugin: string}
}

export const setActiveTelegrafPlugin = (
  telegrafPlugin: string
): SetActiveTelegrafPlugin => ({
  type: 'SET_ACTIVE_TELEGRAF_PLUGIN',
  payload: {telegrafPlugin},
})

interface SetPluginConfiguration {
  type: 'SET_PLUGIN_CONFIGURATION_STATE'
  payload: {telegrafPlugin: TelegrafPluginName}
}

export const setPluginConfiguration = (
  telegrafPlugin: TelegrafPluginName
): SetPluginConfiguration => ({
  type: 'SET_PLUGIN_CONFIGURATION_STATE',
  payload: {telegrafPlugin},
})

interface SetLineProtocolBody {
  type: 'SET_LINE_PROTOCOL_BODY'
  payload: {lineProtocolBody: string}
}

export const setLineProtocolBody = (
  lineProtocolBody: string
): SetLineProtocolBody => ({
  type: 'SET_LINE_PROTOCOL_BODY',
  payload: {lineProtocolBody},
})

interface SetActiveLPTab {
  type: 'SET_ACTIVE_LP_TAB'
  payload: {activeLPTab: LineProtocolTab}
}

export const setActiveLPTab = (
  activeLPTab: LineProtocolTab
): SetActiveLPTab => ({
  type: 'SET_ACTIVE_LP_TAB',
  payload: {activeLPTab},
})

interface SetLPStatus {
  type: 'SET_LP_STATUS'
  payload: {lpStatus: RemoteDataState; lpError: string}
}

export const setLPStatus = (
  lpStatus: RemoteDataState,
  lpError: string = ''
): SetLPStatus => ({
  type: 'SET_LP_STATUS',
  payload: {lpStatus, lpError},
})

interface SetPrecision {
  type: 'SET_PRECISION'
  payload: {precision: WritePrecision}
}

export const setPrecision = (precision: WritePrecision): SetPrecision => ({
  type: 'SET_PRECISION',
  payload: {precision},
})

export const writeLineProtocolAction = (
  org: string,
  bucket: string,
  body: string,
  precision: WritePrecision
) => async dispatch => {
  try {
    dispatch(setLPStatus(RemoteDataState.Loading))

    const resp = await apiPostWrite({
      data: body,
      query: {org, bucket, precision},
    })

    if (resp.status === 204) {
      dispatch(setLPStatus(RemoteDataState.Done))
    } else if (resp.status === 429) {
      dispatch(notify(readWriteCardinalityLimitReached(resp.data.message)))
      dispatch(setLPStatus(RemoteDataState.Error))
    } else {
      throw new Error(get(resp, 'data.message', 'Failed to write data'))
    }
  } catch (error) {
    console.error(error)
    dispatch(setLPStatus(RemoteDataState.Error, error.message))
  }
}

export const saveScraperTarget = () => async (
  dispatch: Dispatch<Action>,
  getState: GetState
) => {
  const {
    dataLoading: {
      dataLoaders: {
        scraperTarget: {url, id, name},
      },
      steps: {bucketID, orgID},
    },
  } = getState()

  try {
    if (id) {
      await client.scrapers.update(id, {url, bucketID})
    } else {
      const newTarget = await client.scrapers.create({
        name,
        type: ScraperTargetRequest.TypeEnum.Prometheus,
        url,
        bucketID,
        orgID,
      })
      dispatch(setScraperTargetID(newTarget.id))
    }
  } catch (error) {
    console.error()
  }
}
