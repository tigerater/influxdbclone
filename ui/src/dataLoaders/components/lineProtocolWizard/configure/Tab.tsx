import React, {PureComponent} from 'react'

import {Radio} from '@influxdata/clockface'

import {LineProtocolTab} from 'src/types'

interface Props {
  active: boolean
  tab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tab, active} = this.props

    return (
      <Radio.Button
        key={tab}
        id={tab}
        titleText={tab}
        value={tab}
        active={active}
        onClick={this.handleClick}
      >
        {tab}
      </Radio.Button>
    )
  }

  private handleClick = () => {
    this.props.onClick(this.props.tab)
  }
}
