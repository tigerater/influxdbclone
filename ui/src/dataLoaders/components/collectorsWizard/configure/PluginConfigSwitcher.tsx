// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

// Components
import PluginConfigForm from 'src/dataLoaders/components/collectorsWizard/configure/PluginConfigForm'
import EmptyDataSourceState from 'src/dataLoaders/components/configureStep/EmptyDataSourceState'

// Utils
import {getConfigFields} from 'src/dataLoaders/utils/pluginConfigs'

// Types
import {TelegrafPlugin, ConfigFields} from 'src/types/dataLoaders'
import {AppState} from 'src/types'
import TelegrafPluginInstructions from 'src/dataLoaders/components/collectorsWizard/configure/TelegrafPluginInstructions'

interface StateProps {
  telegrafPlugins: TelegrafPlugin[]
}

type Props = StateProps

export class PluginConfigSwitcher extends PureComponent<Props> {
  public render() {
    const {telegrafPlugins} = this.props

    if (this.activeTelegrafPlugin) {
      return (
        <PluginConfigForm
          telegrafPlugin={this.activeTelegrafPlugin}
          configFields={this.configFields}
        />
      )
    } else if (!telegrafPlugins || !telegrafPlugins.length) {
      return <EmptyDataSourceState />
    }

    return <TelegrafPluginInstructions />
  }

  private get activeTelegrafPlugin(): TelegrafPlugin {
    const {telegrafPlugins} = this.props
    return telegrafPlugins.find(tp => tp.active)
  }

  private get configFields(): ConfigFields {
    if (this.activeTelegrafPlugin) {
      return getConfigFields(this.activeTelegrafPlugin.name)
    }
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {telegrafPlugins},
  },
}: AppState): StateProps => ({
  telegrafPlugins,
})

export default connect<StateProps>(
  mstp,
  null
)(PluginConfigSwitcher)
