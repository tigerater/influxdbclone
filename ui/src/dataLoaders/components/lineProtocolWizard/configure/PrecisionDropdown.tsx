// Libraries
import React, {PureComponent} from 'react'

// Components
import {Dropdown} from '@influxdata/clockface'

// Types
import {WritePrecision} from '@influxdata/influx'
import {Precision} from 'src/types/dataLoaders'

interface Props {
  setPrecision: (precision: WritePrecision) => void
  precision: WritePrecision
}

const writePrecisions: WritePrecision[] = [
  WritePrecision.Ns,
  WritePrecision.Us,
  WritePrecision.Ms,
  WritePrecision.S,
]

const makePrecisionReadable = {
  [WritePrecision.Ns]: Precision.Nanoseconds,
  [WritePrecision.Us]: Precision.Microseconds,
  [WritePrecision.S]: Precision.Seconds,
  [WritePrecision.Ms]: Precision.Milliseconds,
}

class PrecisionDropdown extends PureComponent<Props> {
  public render() {
    const {setPrecision, precision} = this.props
    return (
      <div className="wizard-step--lp-precision">
        <label>Time Precision</label>
        <Dropdown
          widthPixels={200}
          className="wizard-step--lp-precision"
          button={(active, onClick) => (
            <Dropdown.Button active={active} onClick={onClick}>
              {makePrecisionReadable[precision]}
            </Dropdown.Button>
          )}
          menu={onCollapse => (
            <Dropdown.Menu onCollapse={onCollapse}>
              {writePrecisions.map(value => (
                <Dropdown.Item
                  key={value}
                  value={value}
                  id={value}
                  onClick={setPrecision}
                  selected={`${value}` === `${precision}`}
                >
                  {makePrecisionReadable[value]}
                </Dropdown.Item>
              ))}
            </Dropdown.Menu>
          )}
        />
      </div>
    )
  }
}

export default PrecisionDropdown
