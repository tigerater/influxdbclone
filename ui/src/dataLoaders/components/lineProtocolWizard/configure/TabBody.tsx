// Libraries
import React, {PureComponent, ChangeEvent} from 'react'

// Components
import {
  Form,
  Input,
  Grid,
  TextArea,
  Columns,
  InputType,
  ComponentSize,
} from '@influxdata/clockface'
import DragAndDrop from 'src/shared/components/DragAndDrop'
import {LineProtocolTab} from 'src/types'
import {WritePrecision} from '@influxdata/influx'

import {setLineProtocolBody} from 'src/dataLoaders/actions/dataLoaders'

interface Props {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
  setLineProtocolBody: typeof setLineProtocolBody
  onURLChange: (url: string) => void
  urlInput: string
}

export default class extends PureComponent<Props> {
  public render() {
    const {lineProtocolBody, activeLPTab, urlInput} = this.props

    switch (activeLPTab) {
      case LineProtocolTab.UploadFile:
        return (
          <DragAndDrop
            submitText="Upload File"
            handleSubmit={this.handleSetLineProtocol}
            submitOnDrop={true}
            submitOnUpload={true}
          />
        )
      case LineProtocolTab.EnterManually:
        return (
          <TextArea
            value={lineProtocolBody}
            placeholder="Write text here"
            onChange={this.handleTextChange}
          />
        )
      case LineProtocolTab.EnterURL:
        return (
          <Grid>
            <Grid.Row>
              <Grid.Column
                widthXS={Columns.Twelve}
                widthMD={Columns.Ten}
                offsetMD={Columns.One}
              >
                <Form.Element label="File URL:">
                  <Input
                    titleText="File URL:"
                    type={InputType.Text}
                    placeholder="http://..."
                    value={urlInput}
                    onChange={this.handleChange}
                    autoFocus={true}
                    size={ComponentSize.Large}
                  />
                </Form.Element>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        )
    }
  }

  private handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const {value} = e.target
    this.props.onURLChange(value)
  }

  private handleTextChange = async (e: ChangeEvent<HTMLTextAreaElement>) => {
    const {setLineProtocolBody} = this.props
    setLineProtocolBody(e.target.value)
  }

  private handleSetLineProtocol = (lpBody: string) => {
    const {setLineProtocolBody} = this.props
    setLineProtocolBody(lpBody)
  }
}
