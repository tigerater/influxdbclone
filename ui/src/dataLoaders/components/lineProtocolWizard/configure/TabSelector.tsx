import React, {PureComponent} from 'react'

import {Radio, ButtonShape} from '@influxdata/clockface'
import {LineProtocolTab} from 'src/types'

import Tab from 'src/dataLoaders/components/lineProtocolWizard/configure/Tab'

interface Props {
  tabs: LineProtocolTab[]
  activeLPTab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tabs, activeLPTab} = this.props
    return (
      <Radio shape={ButtonShape.Default}>
        {tabs.map(t => (
          <Tab
            tab={t}
            key={t}
            active={activeLPTab === t}
            onClick={this.handleTabClick}
          />
        ))}
      </Radio>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {activeLPTab, onClick} = this.props
    if (tab !== activeLPTab) {
      onClick(tab)
    }
  }
}
