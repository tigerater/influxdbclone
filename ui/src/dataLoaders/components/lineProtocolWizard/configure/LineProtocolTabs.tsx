// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

// Components
import PrecisionDropdown from 'src/dataLoaders/components/lineProtocolWizard/configure/PrecisionDropdown'
import TabSelector from 'src/dataLoaders/components/lineProtocolWizard/configure/TabSelector'
import TabBody from 'src/dataLoaders/components/lineProtocolWizard/configure/TabBody'

// Types
import {AppState, LineProtocolTab} from 'src/types'
import {WritePrecision} from '@influxdata/influx'

// Actions
import {
  setLineProtocolBody,
  setActiveLPTab,
  setPrecision,
} from 'src/dataLoaders/actions/dataLoaders'

interface OwnProps {
  tabs: LineProtocolTab[]
  bucket: string
  org: string
}

type Props = OwnProps & DispatchProps & StateProps

interface DispatchProps {
  setLineProtocolBody: typeof setLineProtocolBody
  setActiveLPTab: typeof setActiveLPTab
  setPrecision: typeof setPrecision
}

interface StateProps {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
}

interface State {
  urlInput: string
}

export class LineProtocolTabs extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      urlInput: '',
    }
  }

  public render() {
    const {
      setPrecision,
      precision,
      activeLPTab,
      tabs,
      setLineProtocolBody,
      lineProtocolBody,
    } = this.props

    const {urlInput} = this.state

    return (
      <>
        <div className="line-protocol--header">
          <TabSelector
            activeLPTab={activeLPTab}
            tabs={tabs}
            onClick={this.handleTabClick}
          />
          <PrecisionDropdown
            setPrecision={setPrecision}
            precision={precision}
          />
        </div>
        <TabBody
          onURLChange={this.handleURLChange}
          activeLPTab={activeLPTab}
          precision={precision}
          urlInput={urlInput}
          lineProtocolBody={lineProtocolBody}
          setLineProtocolBody={setLineProtocolBody}
        />
      </>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {setActiveLPTab, setLineProtocolBody} = this.props

    setLineProtocolBody('')
    setActiveLPTab(tab)
  }

  private handleURLChange = (urlInput: string) => {
    this.setState({urlInput})
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, activeLPTab, precision},
  },
}: AppState) => {
  return {lineProtocolBody, activeLPTab, precision}
}

const mdtp: DispatchProps = {
  setLineProtocolBody,
  setActiveLPTab,
  setPrecision,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocolTabs)
