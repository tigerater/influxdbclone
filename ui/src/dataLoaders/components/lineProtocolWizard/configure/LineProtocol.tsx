// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'

import _ from 'lodash'

// Components
import {Form} from '@influxdata/clockface'
import LineProtocolTabs from 'src/dataLoaders/components/lineProtocolWizard/configure/LineProtocolTabs'
import OnboardingButtons from 'src/onboarding/components/OnboardingButtons'
import FancyScrollbar from 'src/shared/components/fancy_scrollbar/FancyScrollbar'

// Actions
import {
  setLPStatus as setLPStatusAction,
  writeLineProtocolAction,
} from 'src/dataLoaders/actions/dataLoaders'

// Decorator
import {ErrorHandling} from 'src/shared/decorators/errors'

// Types
import {LineProtocolTab} from 'src/types/dataLoaders'
import {AppState} from 'src/types/index'
import {WritePrecision} from '@influxdata/influx'
import {RemoteDataState} from 'src/types'
import {LineProtocolStepProps} from 'src/dataLoaders/components/lineProtocolWizard/LineProtocolWizard'

type OwnProps = LineProtocolStepProps

interface StateProps {
  lineProtocolBody: string
  precision: WritePrecision
  bucket: string
  org: string
}

interface DispatchProps {
  setLPStatus: typeof setLPStatusAction
  writeLineProtocolAction: typeof writeLineProtocolAction
}

type Props = OwnProps & StateProps & DispatchProps

@ErrorHandling
export class LineProtocol extends PureComponent<Props> {
  public componentDidMount() {
    const {setLPStatus} = this.props
    setLPStatus(RemoteDataState.NotStarted)
  }

  public render() {
    return (
      <div className="onboarding-step">
        <Form onSubmit={this.handleSubmit}>
          <FancyScrollbar
            autoHide={true}
            className="wizard-step--scroll-content"
          >
            <div>
              <h3 className="wizard-step--title">Add Data via Line Protocol</h3>
              <h5 className="wizard-step--lp-sub-title">
                Need help writing InfluxDB Line Protocol?{' '}
                <a
                  href="https://v2.docs.influxdata.com/v2.0/write-data/#write-data-in-the-influxdb-ui"
                  target="_blank"
                >
                  See Documentation
                </a>
              </h5>

              {this.content}
            </div>
          </FancyScrollbar>
          <OnboardingButtons autoFocusNext={true} />
        </Form>
      </div>
    )
  }

  private get LineProtocolTabs(): LineProtocolTab[] {
    return [LineProtocolTab.UploadFile, LineProtocolTab.EnterManually]
  }

  private get content(): JSX.Element {
    const {bucket, org} = this.props
    return (
      <LineProtocolTabs
        tabs={this.LineProtocolTabs}
        bucket={bucket}
        org={org}
      />
    )
  }

  private handleSubmit = () => {
    const {onIncrementCurrentStepIndex} = this.props
    this.handleUpload()
    onIncrementCurrentStepIndex()
  }

  private handleUpload = async () => {
    const {bucket, org, lineProtocolBody, precision} = this.props
    this.props.writeLineProtocolAction(org, bucket, lineProtocolBody, precision)
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, precision},
    steps: {bucket},
  },
  orgs,
}: AppState): StateProps => {
  return {lineProtocolBody, precision, bucket, org: orgs.org.name}
}

const mdtp: DispatchProps = {
  setLPStatus: setLPStatusAction,
  writeLineProtocolAction,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocol)
