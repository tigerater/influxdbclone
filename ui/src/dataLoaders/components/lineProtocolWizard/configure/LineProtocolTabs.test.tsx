// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import {LineProtocolTabs} from 'src/dataLoaders/components/lineProtocolWizard/configure/LineProtocolTabs'

import {LineProtocolTab} from 'src/types/dataLoaders'

const setup = (override?) => {
  const props = {
    tabs: [
      LineProtocolTab.UploadFile,
      LineProtocolTab.EnterManually,
      LineProtocolTab.EnterURL,
    ],
    bucket: 'a',
    org: 'a',
    ...override,
  }

  const wrapper = shallow(<LineProtocolTabs {...props} />)

  return {wrapper}
}

describe('LineProtocolTabs', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)
    })
  })
})
