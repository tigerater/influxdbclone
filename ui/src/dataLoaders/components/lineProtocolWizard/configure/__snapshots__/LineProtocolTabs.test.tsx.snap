// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineProtocolTabs rendering renders! 1`] = `
<div>
  <default_1
    onClick={[Function]}
    tabs={
      Array [
        "Upload File",
        "Enter Manually",
        "Enter URL",
      ]
    }
  />
  <Grid
    testID="grid"
  >
    <GridRow
      testID="grid--row"
    >
      <GridColumn
        offsetLG={2}
        offsetMD={1}
        testID="grid--column"
        widthLG={8}
        widthMD={10}
        widthXS={12}
      >
        <div
          className="onboarding--admin-user-form"
        >
          <div
            className="wizard-step--lp-body"
          >
            <PrecisionDropdown />
            <default_1
              onURLChange={[Function]}
              urlInput=""
            />
          </div>
        </div>
      </GridColumn>
    </GridRow>
  </Grid>
</div>
`;
